src/channel_manager.ts

```diff
@@ -0,0 +1,527 @@
+  | 'notificationNewMessageHandler'
+  | 'notificationRemovedFromChannelHandler';
+
+export type ChannelManagerEventHandlerOverrides<SCG extends ExtendableGenerics = DefaultGenerics> = Partial<
+  Record<ChannelManagerEventHandlerNames, EventHandlerOverrideType<SCG>>
+>;
+
+export const channelManagerEventToHandlerMapping: {
+  [key in ChannelManagerEventTypes]: ChannelManagerEventHandlerNames;
+} = {
+  'channel.deleted': 'channelDeletedHandler',
+  'channel.hidden': 'channelHiddenHandler',
+  'channel.truncated': 'channelTruncatedHandler',
+  'channel.visible': 'channelVisibleHandler',
+  'message.new': 'newMessageHandler',
+  'member.updated': 'memberUpdatedHandler',
+  'notification.added_to_channel': 'notificationAddedToChannelHandler',
+  'notification.message_new': 'notificationNewMessageHandler',
+  'notification.removed_from_channel': 'notificationRemovedFromChannelHandler',
+};
+
+export type ChannelManagerOptions = {
+  allowNewMessagesFromUnfilteredChannels?: boolean;
+  lockChannelOrder?: boolean;
+};
+
+export const DEFAULT_CHANNEL_MANAGER_OPTIONS = {
+  allowNewMessagesFromUnfilteredChannels: true,
+  lockChannelOrder: false,
+};
+
+export class ChannelManager<SCG extends ExtendableGenerics = DefaultGenerics> {
+  public readonly state: StateStore<ChannelManagerState<SCG>>;
+  private client: StreamChat<SCG>;
+  private unsubscribeFunctions: Set<() => void> = new Set();
+  private eventHandlers: Map<string, EventHandlerType<SCG>> = new Map();
+  private eventHandlerOverrides: Map<string, EventHandlerOverrideType<SCG>> = new Map();
+  private options: ChannelManagerOptions = {};
+  private stateOptions: ChannelStateOptions = {};
+
+  constructor({
+    client,
+    eventHandlerOverrides = {},
+    options = {},
+  }: {
+    client: StreamChat<SCG>;
+    eventHandlerOverrides?: ChannelManagerEventHandlerOverrides<SCG>;
+    options?: ChannelManagerOptions;
+  }) {
+    this.client = client;
+    this.state = new StateStore<ChannelManagerState<SCG>>({
+      channels: [],
+      pagination: {
+        isLoading: false,
+        isLoadingNext: false,
+        hasNext: false,
+        filters: {},
+        sort: {},
+        options: { limit: 10, offset: 0 },
+      },
+      ready: false,
+    });
+    this.setEventHandlerOverrides(eventHandlerOverrides);
+    this.setOptions(options);
+    this.eventHandlers = new Map(
+      Object.entries({
+        channelDeletedHandler: this.channelDeletedHandler,
+        channelHiddenHandler: this.channelHiddenHandler,
+        channelVisibleHandler: this.channelVisibleHandler,
+        memberUpdatedHandler: this.memberUpdatedHandler,
+        newMessageHandler: this.newMessageHandler,
+        notificationAddedToChannelHandler: this.notificationAddedToChannelHandler,
+        notificationNewMessageHandler: this.notificationNewMessageHandler,
+        notificationRemovedFromChannelHandler: this.notificationRemovedFromChannelHandler,
+      }) as [string, EventHandlerType<SCG>][],
```

Comment:
```suggestion
      Object.entries<EventHandlerType<SCG>>({
        channelDeletedHandler: this.channelDeletedHandler,
        channelHiddenHandler: this.channelHiddenHandler,
        channelVisibleHandler: this.channelVisibleHandler,
        memberUpdatedHandler: this.memberUpdatedHandler,
        newMessageHandler: this.newMessageHandler,
        notificationAddedToChannelHandler: this.notificationAddedToChannelHandler,
        notificationNewMessageHandler: this.notificationNewMessageHandler,
        notificationRemovedFromChannelHandler: this.notificationRemovedFromChannelHandler,
      }),
```
This then reveals that `EventTypeHandler` is not properly taking into account all the possible handler return combinations (in this case `Promise<void>`) - maybe we should just do `any` - it's not like we really care about the outputs, right?

Reply:
Sure ! We can do that. While it is true that we don't particularly care about what it returns (it should generally be void, though) I thought it'd make sense to cover what's expected - hence the typing. The general idea was to avoid the expectation that this also returns a value.